weaknesses/MASVS-CRYPTO/MASWE-0021.md

```diff
@@ -11,15 +11,45 @@ mappings:
+
+Any algorithm not meeting the above conditions will be considered risky or too "weak" for general use in hashing. When a collision attack is discovered and is found to be faster than a birthday attack, a hash function is often denounced as "broken". This is the case for MD5 and SHA-1.
+
+Another common issue is using an HKDF for key derivation with any type of integrity-based hashing algorithm like MD5, SHA-1, SHA-2, or even SHA-3 on low-entropy input like user-supplied passwords and pins. HKDF isn't designed for low-entropy inputs. Doing so will produce "weak" hashes that can easily be broken.
+
+## Impact
+
+- **Loss of authenticity**: A hashing algorithm, known to be vulnerable to collision attacks, may compromise the authenticity of the data as it allows for two data sources to be identical.
+- **Loss of integrity**: Algorithms that are susceptible to length extension attacks may allow an attacker to compromise the integrity of the data by appending data to the original data source.
+- **Loss of confidentiality**: A hashing algorithm, known to be vulnerable to pre-image attacks, increases the likelihood that encrypted data can be leaked through using cryptoanalysis or brute-force.
+
+## Modes of Introduction
+
+- **Using a deprecated, risky, or broken hashing algorithm**: E.g., MD5 and SHA-1 have been identified to be vulnerable to collision attacks that are faster than a birthday attack.
+- **Using a hash susceptible to length extension attacks**: E.g., MD5 and SHA-1 have been identified to be vulnerable to length extension attacks.
+- **Using non-resource-intensive algorithms on low-entropy input**: Using an integrity-based hashing algorithm to hash low-entropy input like pin numbers would make brute-force or dictionary attacks trivial.
+
+## Mitigations
+
+- **Choose collision-resistant algorithm**: Choose an algorithm that is sufficiently collision-resistant, like the integrity algorithms SHA-2 (with 256, 384, 512 bits), BLAKE3 and the SHA-3 family
+- **Choose an algorithm with sufficient bit-lengths**: As our computers get stronger, the hashes get weaker, therefore, make sure that you can adjust the bit-length of the algorithm of your choosing. When hashes are stored at rest, make sure to follow the software industry's long-term recommendations (e.g., ["NIST: Transition to Post-Quantum Cryptography Standards](https://csrc.nist.gov/pubs/ir/8547/ipd)").
+- **Choose an algorithm fit for its purpose**: To ensure the data's integrity, choose an integrity-based algorithm. When you want to hash low-entropy input, choose a password hash algorithm. Don't try to be clever. Follow recommendations and guidelines.
```

Comment:
I cannot find BLAKE3 support in Android/iOS therefore I wouldn't include at this level, even if it's correct.

```suggestion
- **Choose collision-resistant algorithm with sufficient bit-lengths**: Choose SHA-2 (256, 384, or 512 bits) or the SHA-3 family for integrity and fingerprinting purposes.
- **Match the algorithm to the purpose**: Use password hashing functions for passwords and PINs. Use hash based KDFs only when the input secret already has high entropy. Do not treat general purpose hash functions as password hashing or key stretching mechanisms.
- **Avoid truncating hashes too aggressively**: If you truncate a hash, keep enough bits for the security level you need, for example 128 bits or more for long term integrity.
```